---
title: "RNASeqProj"
author: "Gavin Fortenberry"
date: "July 16, 2018"
output: html_document
---
```{r}
install.packages("dplyr")
#install.packages("questionr")
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown
## Reads the two CSV's and combines them into one data frame:
```{r}

list.files("../RNASeqData")
df_pheno <- read.csv(file="../RNASeqData/phenotypeTable.csv", header =TRUE, sep=",")
df_molecular <- read.csv(file="../RNASeqData/molecularDataSets.csv", header =TRUE, sep=",")
#df_pheno
#df_molecular
df_pheno <- dplyr::rename(df_pheno, assay_material_id = `ï..assay_material_id`)
#str(df_pheno)
#str
df_combined <- dplyr::full_join(df_pheno, df_molecular, by = "assay_material_id")
df_combined
```

## Writes a CSV from the new combined data frame:
```{r}
write.csv(df_combined, file = "../ProjectFiles/combineddata.csv", row.names = FALSE)
```

##Reads in the commbined data set/tests structure
```{r}
test_set_combined_data <- read.csv("combineddata.csv")
str(test_set_combined_data)
summary(test_set_combined_data)
```

##Make list of directories(folders) 
```{r}
#":/Users/gfortenb/Documents/GitHub/bioDS-bootcamp/RNASeqData"
RNADirectoryList = list.dirs( path = "C:/Users/gfortenb/Documents/GitHub/bioDS-bootcamp/RNASeqData", 
                              
                              recursive = FALSE)

```
##Make List of files within each folder
```{r}
RNADirectoryList
FileList1 = sapply(RNADirectoryList, 
                   function(x){list.files(path = x, 
                                          full.names = TRUE, 
                                          pattern = "htseq.txt") }) #makes a list of files with that string in the name
FileIDList <- as.data.frame(cbind(FileList1, gsub("^.*-","", RNADirectoryList)), stringsAsFactors = F)
colnames(FileIDList) <- c("Path", "molecular_id")
FileList1
```
##Read in data from selected list and compile into data frame

```{r}
listOf_all_df <- lapply(seq(1:nrow(FileIDList)),
                 function(i){ 
                   X <-read.delim(file = FileIDList$Path[i],
                                         header = FALSE);
                   colnames(X) <- c("Gene", FileIDList$molecular_id[i]);
                   return(X)
                   } )

listOf_all_df
#make list of dataframes into one dataframe w/all contents of each dataframe in the list of dataframes as columns

```

