---
title: "RNA Sequencing Differential Analysis Project"
author: "Gavin Fortenberry"
date: "August 14th, 2018"
output:
  pdf_document:
    toc: yes
  html_document:
    toc: yes
    toc_float: yes
  word_document:
    toc: yes
---
# Introduction

### What is Myleofibrosis?
- Myleofibrosis is a type of bone marrow cancer in which an rapidly increasing number of blood forming cells form a fibrous like structure that sometimes leads to acute leukemia. 
Certain genotypes like the JAK2 V617F mutation have been a determining factor of Blast Transformation in Myleofibrosis.

## What is differential gene expression?
- Differential gene expression is a way to analyze factors in different groups that may or may not be associated with different gene counts, from RNA sequence data. 

### What is gene expression?
- Gene expression is the process of which information inside a gene is used to make RNA and proteins. 
The genotype leads to the phenotype.

### Why is differential gene expression important to a biological question?



##How to do differential gene expression:


### What inputs go in:
24 patients, 12 MF, 12 normal
### What is the analysis going to do

### What are the outputs/what is the meaning of the outputs



##(My) Experimental Design

### What groups am I going to compare? (Bio factors vs. tech factors)


- The groups of phenotype data being used can be sorted into two overarching categories: biological factors and technical factors.
Biological factors include "Tissue_type", "genotype_jak2", "genotype_calr", and "genotype_mpl".
Technical factors include: "collection_type", "time_to_processing", and "extraction_type".

- The reason why I am splitting the data/analysis into two categories is to attempt to see whether certain factors from each category have an impact on specific gene counts or not.

- To be completed: Why im picking each of the columns within those sections

- I will separate demographic groups by phenotype, and compare gene expression counts for each of the genes between the
two groups, groups defined by conditions of that factor. For example, the conditions of the factor of age could be different age groups, sex could be male or female, genotype could be present or not present, etc.
ultimate goal is to make observations/conclusions about possible differences in gene counts between different groups that may or may not be significant in diagnosis of Myleofibrosis.

# Data Import

### Install R Packages
Installs packages of functions that can be used to help manipulate variables and data libraries. These packages can be installed through base R code that downloads them and installs them into R studio. Then certain programming phrases can be used to do new functions in connection with base R programming phrases, with an indicator of the package being used mentioned before using that package's functions. For example, if I wanted to use a function of the "dplyr" package, I would write -> "dplyr::(insert function here)(insert arguments of functions here)". Note that functions or phrases that are in base R programming vocabulary do not require a name indicator of the package being used, as it is from the base software and the software knows how to interperet those phrases without an idicator of a package. It is still required to write names of functions from Base R that are wanting to be used though, and if shortcut variables for functions are desired then you can assign new variables to those functions.

Bioconductor packages need to be installed by "biocLite" rather than install.packages which is for CRAN/base R
```{r Install Packages, eval=FALSE, message=FALSE, warning=FALSE, include=TRUE}
install.packages("dplyr")
install.packages("knitr")
install.packages("rmarkdown")




### Install Bioconductor, the DESeq2 Package and ggplot2
source("https://bioconductor.org/biocLite.R")
biocLite("DESeq2")
install.packages("ggplot2")
```

### Load Additional Packages
```{r Load Additional Libraries, eval=TRUE, message=FALSE, warning=FALSE, include=TRUE}
library(DESeq2); library(ggplot2)
library(plyr); library(dplyr)
library(gghighlight)
```

### Read in Project Metadata to R
Reads the two CSV's and combines them into one data frame:
The code below creates three data frames: one labeled as "df_pheno" for data of a  file called phenotable, which contains demographics and general information from patients and samples taken from them, another labeled as "df_molecular" for data from another file called molecularDataSets, which contains more specific info on samples collected from patients and their diagnoses from these samples, and another labeled as "df_combined" that combines the two data frames according to common columns.

The overall purpose of this code chunk is to create the combined data frame.

The first two data frames created ("df_pheno" and "df_molecular") were created using base R code "read.csv" function, which reads in data into a data frame from a CSV, a comma seperated values document. 

In the process of formatting the df_pheno dataframe in order to be compatible when joining with the df_pheno dataframe, the "rename" function from the dplyr package is used on the df_pheno dataframe to change the name of a mismatching column name.

The function used to actually combine the two dataframes into one is the "full_join" function from the dplyr package which joins two indicated dataframes together by a common column.

"str()" is short for structure, and is a base R function that allows the user to get an idea of the format of the data they are looking at
```{r First Data Frame Creations/Combine, echo=TRUE, warning=FALSE}

df_pheno <- read.csv(file="../RNASeqData/phenotypeTable.csv", 
                     header =TRUE)
df_pheno <- df_pheno %>% rename(assay_material_id = ends_with("assay_material_id"))

df_molecular <- read.csv(file="../RNASeqData/molecularDataSets.csv", 
                         header =TRUE, sep=",")

df_combined <- dplyr::full_join(df_pheno, df_molecular, 
                                by = "assay_material_id")
head(df_combined, n = 3)

```

## Save Prepared Metadata
Writes a CSV (a comma seperated values document) from the new combined data frame(df_combined) of dataframes df_pheno and df_molecular into a document called "combineddata.csv" stored in the working directory (the main location of the files created from the Rstudio
```{r Write First CSV, eval=FALSE, include=TRUE}
write.csv(df_combined, file = "../ProjectFiles/combineddata.csv", 
          row.names = FALSE)
```

The read.csv function is used to re-read the newly written CSV to make sure the data is the same as it was when written. The "str()" and "summary" functions are used to compare the statistics of the new dataframe to the original created one to confirm similarity.
```{r Test Output of newly written CSV, eval=FALSE, include=TRUE}
test_set_combined_data <- read.csv("combineddata.csv")
str(test_set_combined_data)
```

## Read in RNA Sequencing Data from HiSat2/htseqcount
Makes a list of directories(folders) with the R base function "list.dirs" which takes the indicated path of the main directory containing the directories its making a list of in, and the econd argument written "recursive" is set to false becuse the main directory is not desired to be listed in the list of its components.

```{r Creation of List of Directories}
#":/Users/gfortenb/Documents/GitHub/bioDS-bootcamp/RNASeqData"
RNADirectoryList = list.dirs(path = "../RNASeqData", recursive = FALSE)

```

Makes a List of files within each folder of each directory in the main directory. Uses a function to go through the files within each folder and only list files with a certain phrase in the name of the file, "htseq.txt", using the pattern function (only argument usedin function is name of character phrase its looking for).

Binds path's of files(locations of the files in the computer) on the list made to the molecular ID of the files with "as.data.frame"" function"/"cbind" function, created by finding key character sequences in the title of the folders containing the "htseq.txt"" files, using the gsub function (from base R). The first argument used in the gsub function is a character phrase that indicates where in the string to look for the character phrase of the molecular ID, and the second argument is the list of data of file locations/names to look for the ID in.

The "colnames" function (base R) sets the column names if the newly created dataframe.

```{r List of htseq files & Path/ID Dataframe, echo=TRUE}
FileList_htseq1 = sapply(RNADirectoryList, 
                   function(x){list.files(path = x, 
                                          full.names = TRUE, 
                                          pattern = "htseq.txt") }) 

htseq_FileID_df <- as.data.frame(cbind(FileList_htseq1, 
                                  gsub("^.*-","", RNADirectoryList)), 
                            stringsAsFactors = F)
colnames(htseq_FileID_df) <- c("Path", "molecular_id")
head(htseq_FileID_df, n = 3)
```
Reads all data from list of selected files and compiles into different data frames/list of different data frames using LApply function.

```{r Creates List of all Dataframes}
listOf_alldf <- lapply(seq(1:nrow(htseq_FileID_df)),
                 function(i){ 
                   X <- read.delim(file = htseq_FileID_df$Path[i],
                                         header = FALSE);
                   colnames(X) <- c("Gene", htseq_FileID_df$molecular_id[i]);
                   return(X)
                   } )
```








###SALMON HERE<<<---
```{r Reads in salmon count data/make DF}
### Reads in data/creates salmon counts dataframe
salmonCounts_df <-read.csv(file="../RNASeqData/2018-08-06_SalmonGeneLevelCounts.csv", 
                     header =TRUE)
```
# Data Summarizing
##Summarizing PHENOTYPES/defining experimental design

###Biological factors
```{r echo=TRUE}
bio_factors_summary <- df_pheno %>% group_by(diagnosis, genotype_jak2, genotype_calr, age_range, sex) %>% summarise(n())
bio_factors_summary
```

###Technological factors
```{r echo=TRUE}
tech_factors_summary <- df_pheno %>% group_by(time_to_processing, collection_type, collection_event, extraction_type) %>% summarise(n())
tech_factors_summary
```

###Bio & tech factors
```{r}
bio_and_tech_summary <- df_pheno %>% group_by(diagnosis, time_to_processing, genotype_jak2, collection_type,  genotype_calr, collection_event, age_range, extraction_type, sex) %>% summarise(n())
bio_and_tech_summary
```




### Create a Summarized Experiment Data set
("Summarized Experiment" - something specific to DESeq2 package)
- Normalize RNA Sequencing Counts - in new normalized data frame
- (Normalize each sample's counts data based on over all library size for each sample.)

# Data Setup For Analysis
## Differential Gene Expression Analysis with DESeq2

####HERE DOWN->

Make the list of dataframes into one dataframe w/all contents of each dataframe in the dataframe of dataframes as columns using the "join_all" function from the plyr package.
This produces a dataframe with the molecular ID and Genes in each sample assigned to a molecular ID columns.
```{r Join All HTSEQ count data, message=FALSE, warning=FALSE}
###THIS MATTERS- MOLECULAR ID's LINKED TO THE PHENOTABLE
htseq_genecounts_df <- plyr:: join_all(listOf_alldf, by = NULL, 
                                 type = "full", match = "all")
head(htseq_genecounts_df, n = 3)
```

Optimizing compatability for creating DataSets for DESEQ2 1(EDITS = DONE)
```{r Optimizing compatability of data for creating DataSets for DESEQ2 - HTSEQ Matrix (part 1)}

#Creats HTSEQ counts matrix
#Imports everything from HTSEQ Counts dataframe except the first column (Doesn't delete column from original dataframe)
htseqCountsMat = as.matrix(htseq_genecounts_df[,-1]); ncol(htseqCountsMat)          #COLS
head(htseqCountsMat)

#Sets rownames of new matrix of HTSEQ count data to gene names in 1st column of original dataframe
rownames(htseqCountsMat)<- htseq_genecounts_df[,1] 




#Repeated processes for SALMON
```
##Repeated processes w/salmon.txt files
```{r }

# Assigns gene names from salmon counts dataframe to variable in order to keep them set as characters/strings
salmonCounts_genes <- as.character(salmonCounts_df[ , 25] )
# Creates matrix of salmon counts from salmon count data frame
# Converts numbers with decimals to integers (no decimals)  in salmon counts matrix, 
salmonCountsMat <- sapply(salmonCounts_df[ , -25],as.integer) 
#head(salmonCounts_df)
# Sets rownames of new salmon counts matrix to gene names of salmon counts dataframe
rownames(salmonCountsMat) <- salmonCounts_genes
head(salmonCountsMat, n = 4)
str(salmonCountsMat)
```

```{r Optimizing compatability of data for creating DataSets for DESEQ2 - PHENO. Matrix (part 2)}
#Creates matrix phenoMat from df_combined dataframe
# Imports all data/columns except molecular ID column (doesnt delete from original dataframe, just doesnt include on import)
phenoMat = as.matrix(subset(df_combined, select=-molecular_id)); nrow(phenoMat) 

#Sets rownames of phenoMat to data from df_combined  column of molecular ID's
rownames(phenoMat)<- df_combined$molecular_id
```

```{r Optimizing compatability of data for creating DataSets for DESEQ2 - ORDER/ALLIGNING (part 3)}
#Link column names AND column values  of htseqCountsMat to phenoMat row names in same order
phenoMat <- phenoMat[match(colnames(htseqCountsMat),row.names(phenoMat)),]
head(phenoMat)
```
##Creating original DeseqDatasets
(EDITS = DONE)
```{r Make HTSEQ DeseqDataset 1} 
#DataSet for DESEQ from HTSEQ Matrix created 
# variable comparing is diagnosis [levels  = mf (myleofibrosis) and normal (no myleofibrosis) ]
dseq_set_htseq <- DESeqDataSetFromMatrix(htseqCountsMat,phenoMat, design = ~ diagnosis)

# with base level being at "normal" (not myleofibrosis)
dseq_set_htseq$diagnosis <- relevel(dseq_set_htseq$diagnosis, "normal")

#gets rid of rows of gene counts of 1's and 0's
dseq_set_htseq <- dseq_set_htseq[ rowSums(counts(dseq_set_htseq)) > 1, ] 
head(dseq_set_htseq)
```

```{r Creating Salmon Deseq Dataset}

dseq_set_salmon <- DESeqDataSetFromMatrix(salmonCountsMat,phenoMat, design = ~ diagnosis)




dseq_set_salmon$diagnosis <- relevel(dseq_set_salmon$diagnosis, "normal")

dseq_set_salmon <- dseq_set_salmon[ rowSums(counts(dseq_set_salmon)) > 1, ] #gets rid of 1's and 0's
head(dseq_set_salmon)

```
##Creating copies of original datasets for analysis/design formulas
```{r Setup of HTSEQDesign 1}


###base lvl = diagnosis
#Make copy of htseq data set so original is not modified, multiple formulas can be applied
dseq_set_htseq_copy <- dseq_set_htseq
design(dseq_set_htseq_copy) <- formula(~ diagnosis)
dseq_set_htseq_copy <- DESeq(dseq_set_htseq_copy)


```


```{r Setup of HTSEQ Design 2}
###base lvl = age range
dseq_set_htseq_copy2 <- dseq_set_htseq
design(dseq_set_htseq_copy2) <- formula(~ age_range)
dseq_set_htseq_copy2 <- DESeq(dseq_set_htseq_copy2)


```


```{r Setup of Salmon Design 1}
dseq_set_salmon_copy <- dseq_set_salmon
design(dseq_set_salmon_copy ) <- formula(~ diagnosis)
dseq_set_salmon_copy  <- DESeq(dseq_set_salmon_copy )


```



# Analysis


###use Deseqdataset to do analysis using deseq2 tools???

### Following paragraph/description Copied from https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#differential-expression-analysis ...IMPORTANT!~!!!!!!!~!~!~!

###A DESeqDataSet object must have an associated design formula. The design formula expresses the variables which will be used in modeling. The formula should be a tilde (~) followed by the variables with plus signs between them (it will be coerced into an formula if it is not already). The design can be changed later, however then all differential analysis steps should be repeated, as the design formula is used to estimate the dispersions and to estimate the log2 fold changes of the model.

###Note: In order to benefit from the default settings of the package, you should put the variable of interest at the end of the formula and make sure the control level is the first level.


Gives results of HTSEQ design formula 1:
```{r HTSEQ Design 1 Results}
resDseqCopy <- results(dseq_set_htseq_copy)
head(resDseqCopy)
```


Gives results of HTSEQ design formula 2:
```{r Design 2 Results}
resDseqCopy2 <- results(dseq_set_htseq_copy2)
head(resDseqCopy2)
###p val dif = low, chance there is a sig. dif.. threshhold = 0.05?
```

``` {r Salmon Design 1 Results}

res_dseq_set_salmon_copy <- results(dseq_set_salmon_copy)
head(res_dseq_set_salmon_copy)

```



# Prep for plotting visual analysis



##Log2 fold change

###HTSEQ
```{r Plotting l2fc for HTSEQ prep}


###Extract log2FoldChange columns from htseq and salmon datasets,
###merge into combined DF with genes as rownames, use geom-point to plot in GGplot
### x= LF2C Salmon, Y= LF2C HTSEQ



###resDseqCopy vs. res_dseq_set_salmon_copy =diagnosis base level = normal

###head(resDseqCopy)
###head(res_dseq_set_salmon_copy)


###convert rownames to column name gene in both df's
###full join by gene


###creates two new df's with l2fc data

###create new column with values of rownames
resDseqCopy$Gene <- rownames(resDseqCopy)

head(resDseqCopy)

###Subset/put in a copy of the gene column and logfold2 column from the htseq data into new object/df
gene_l2fc_htseq1 <- as.data.frame(resDseqCopy[ ,c(7, 2)])

###Getting rid of rownames (gene list is in new column)
rownames(gene_l2fc_htseq1) = NULL
###Repeating same process for Salmon as HTSEQ
```
###SALMON

```{r Plotting l2fc for Salmon prep}
res_dseq_set_salmon_copy$Gene <- rownames(res_dseq_set_salmon_copy) #salmon
gene_l2fc_salmon1 <- as.data.frame(res_dseq_set_salmon_copy[ ,c(7, 2)])
rownames(gene_l2fc_salmon1) = NULL

```
###Tests
``` {r Plotting l2fc prep tests}
###Tests
head(gene_l2fc_htseq1)
str(gene_l2fc_htseq1)
head(gene_l2fc_salmon1)
str(gene_l2fc_salmon1)
```
###HTSEQ & Salmon
``` {r Plotting l2fc prep - joining/setting}

###Joining two DF's of LF2C by Gene
gene_l2fc_htseqSalmon1_combo <- full_join(gene_l2fc_htseq1, gene_l2fc_salmon1, by = "Gene")

#Setting Column names
colnames(gene_l2fc_htseqSalmon1_combo) <- c("Gene", "L2FC_HTSEQ","L2FC_SALMON")

head(gene_l2fc_htseqSalmon1_combo)
```

##P-value
###HTSEQ


```{r Plotting Pvalue prep htseq}
#resDseqCopy #htseq

#gene_l2fc_htseq1_all <- as.data.frame(resDseqCopy)



#res_dseq_set_salmon_copy #Salmon


gene_pvalue_htseq1 <- as.data.frame(resDseqCopy[ ,c(7, 5)])
###Getting rid of rownames (gene list is in new column)
rownames(gene_pvalue_htseq1) = NULL
```
###Salmon

```{r Plotting Pvalue prep salmon}
###Repeating same process for Salmon as HTSEQ
gene_pvalue_salmon1 <- as.data.frame(res_dseq_set_salmon_copy[ ,c(7, 5)])
rownames(gene_pvalue_salmon1) = NULL
```
###Tests

```{r Plotting Pvalue prep Tests}
###Tests
head(gene_pvalue_htseq1)
str(gene_pvalue_htseq1)

head(gene_pvalue_salmon1)
str(gene_pvalue_salmon1)
```
###HTSEQ & Salmon
```{r Plotting Pvalue prep HTSEQ/SALMON combo}

###Joining two DF's of LF2C by Gene
gene_pvalue_htseqSalmon1_combo <- full_join(gene_pvalue_htseq1, gene_pvalue_salmon1, by = "Gene")

#Setting Column names
colnames(gene_pvalue_htseqSalmon1_combo) <- c("Gene", "PVAL_HTSEQ","PVAL_SALMON")

head(gene_pvalue_htseqSalmon1_combo)

```



# Visualizing analysis/plotting

##Log2 Fold Change and PVAL Plots
```{r Plotting l2fc - part 1 }
p1 = ggplot(data = gene_l2fc_htseqSalmon1_combo, aes(x= L2FC_SALMON, y = L2FC_HTSEQ)) + geom_point()
p1

#EXPLAIN WHAT L2FC IS
#each dot corresponds to 2 logfoldchanges for one gene
#numbers are l2fc of mf vs normal gene expression

```


PVAL Htseq/salmon for diagnosis mf vs. normal
```{r Plotting Pvalue }
p2 = ggplot(data = gene_pvalue_htseqSalmon1_combo, aes(x = PVAL_SALMON, y = PVAL_HTSEQ)) + geom_point()
p2
#geom_boxplot
```
##Box Plots
```{r Box Plot 1}
#boxplot


# htseq then salmon - 
# boxplot of gene expression of gene x based on counts
#for mf, 
#for normal

#Change data frame/matrix to graph from
box1 = ggplot(data = gene_l2fc_htseqSalmon1_combo, aes(x= L2FC_SALMON, y = L2FC_HTSEQ)) + geom_boxplot()
box1
```


```{r Box Plot 2}
#box2 = ggplot(data = gene_pvalue_htseqSalmon1_combo, aes(x= PVAL_HTSEQ, y = PVAL_SALMON)) + geom_boxplot()
#box2
```






##Exploratory plots
###Exploratory plot1

```{r dataset1 gene count plots}
DESeq2::plotCounts(dseq_set_htseq, "CALR", intgroup = "diagnosis")
DESeq2::plotCounts(dseq_set_htseq, "JAK2", intgroup = "diagnosis")
DESeq2::plotCounts(dseq_set_htseq, "A1BG", intgroup = "diagnosis")
DESeq2::plotCounts(dseq_set_htseq, "ADAMTS7", intgroup = "diagnosis")

```




