---
title: "RNAseq Differential Gene Expression Analysis"
output: 
  html_document: 
    fig_caption: yes
    toc: yes
---

# Instructions for using a notebook 

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. Execute a code chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Cmd+Shift+Enter*. The results will appear beneath the code. Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Cmd+Option+I*.

```{r example plot, eval=FALSE, include=FALSE}
# create plot 
plot(cars)
```

# Install and load software 

If you do not already have Tidyverse and DESeq2 installed, this code will perform the task for you. You may be asked the following questions:

* "Update all/some/none"? type a (for all) and hit Enter
* "Do you want to install from sources the package which needs compilation? (Yes/no/cancel)" hit Enter 

```{r eval=FALSE, message=FALSE, warning=TRUE, include=FALSE}
# install tools for data manipulation and plotting
install.packages("tidyverse")

# install tools for genomic analysis
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install("DESeq2")
```

Note that you will need to *install* a package only once per computer, but you'll need to *load* the packages every time you reopen RStudio.

## Load Libraries
After you have installed the packages you need, in a session you will need to load them as libraries so that you can use the functions in packages in your code.  
```{r Load Libraries, message=FALSE, warning=FALSE}
library(tidyverse)
library(DESeq2)
```



# Preparing Your Data

## Importing Metadata
In this repository, there is a folder called "Datasets/RNASeqData" that contains two csv files containing metadata relevant to the study you are performing here.  

The working directory of this project is the RNASeqProject folder.  However, the path to the metadata files, relative to the current working directory is "../Datasets/RNASeqData/phenotypeTable.csv" and "../Datasets/RNASeqData/molecularDataSets.csv". When working with your own data, be sure to save the data files in a place where you can find this path for your own work and replace them with what we have here.  

Here we will read these files in to R and store them as data frames.  
```{r}
molecularMeta <- read.csv(file = "../Datasets/RNASeqData/molecularDataSets.csv", 
                          # this is the path and filename of the file to read in
                          header = T, 
                          # the file has a header row, so we want that to become the column names of the data frame
                          stringsAsFactors = F
                          # Advancexd topics:  read about strings vs factors!!
                          )

phenotypeMeta <- read.csv(file = "../Datasets/RNASeqData/phenotypeTable.csv", 
                          # this is the path and filename of the file to read in
                          header = T, 
                          # the file has a header row, so we want that to become the column names of the data frame
                          stringsAsFactors = F
                          # Advancexd topics:  read about strings vs factors!!
                          )

```

## Using tidyverse functions to learn about a dataset
The "tidyverse" is a group of R packages that work together to form a handy language for manipulating datasets in R.  While it contains a wide array of functionalities, we are going to primarily focus here on the use of the `dplyr` package for manipulating data frames.  

### Get a feel for your metadata
First you must get a feeling for what the data actually contain.  Here are some starting points for you to explore our two data frames.  

```{r echo=TRUE}
# What are the names of our columns?
colnames(molecularMeta)
colnames(phenotypeMeta)

# Are there columns that are similar in the two data frames?
colnames(molecularMeta) %in% colnames(phenotypeMeta)

# If so, which ones?
colnames(molecularMeta)[colnames(molecularMeta) %in% colnames(phenotypeMeta)]

# Can we make one large dataset that contains both the phenotype and the molecular metadata?
studyMetadata <- dplyr::full_join(phenotypeMeta, molecularMeta)
studyMetadata <- full_join(phenotypeMeta, molecularMeta) # Equivalent to the line above

```

Let's save this file so we can use/view it outside of R
```{r eval=FALSE, include=FALSE}
write.csv(x = studyMetadata, # the data frame you want to save
          file = "Results/overallStudyMetadata.csv", # the path/filename you want to save it to
          quote = F, # rarely does anyone love quotes around the text!
          row.names = F) # row names are usually silly until they aren't
```

In the future we could just skip the above and read in the saved data directly
```{r eval=FALSE, include=FALSE}
studyMetadata <- read.csv(file = "Results/overallStudyMetadata.csv",
                          header = T,
                          stringsAsFactors = F)

```



### How many different values are possible for the field `diagnosis` in our study?
```{r}
studyMetadata %>% group_by(diagnosis) %>% summarize(numberOfSamples = n())
```


### What are possible confounding variables in our study?
```{r}
studyMetadata %>% group_by(diagnosis, genotype_jak2, genotype_calr, rin_range) %>% summarize(numberOfSamples = n())
```

In this case, we see that all of the normal blood samples had high quality RNA (`rin_range`), but the myelofibrosis samples had more mixed qualities.  We can look at the groups of `rin_range` more specifically to see how many samples are in each group.  

```{r}
studyMetadata %>% group_by(diagnosis, rin_range) %>% summarize(numberOfSamples = n())
```

Also, in the myelofibrosis samples, genotyping results (where the sample is tested to see if the DNA has a specific mutation or not in a gene) show that some of the patients had a mutation in JAK2, while some had a mutation in CALR.  Let's see specifically how many samples fall into each group based on genotyping results.  
```{r}
studyMetadata %>% group_by(diagnosis, genotype_jak2, genotype_calr) %>% summarize(numberOfSamples = n())
```


In both cases, there is one sample that doesn't fit into any group, and either 5 or 6 samples in other `rin_range` or `genotype` groups.  While these variables are not the primary variables we are trying to study using these RNA sequencing results, we need to keep in mind that they might have an impact on our results that might obscure the effect we are trying to measure.  


## Importing RNA Seq Data
Now we need to read in the processed results from our RNA sequencing data analysis.  In this example we are going to read in teh files created by a bioinformatic workflow applied to the raw RNA sequencing results generated by the sequencer.  The raw data contain the individual sequences of the fragments of DNA that were created from the RNA and sequenced.  The workflow used `HiSat2` to align these reads to the human genome reference and then `htseq` was used to count how many of these reads were aligned to regions of the genome we consider part of individual genes.  Thus, the results of this bioinformatic process create a table of data for each of the biospecimens that contain a column with Gene IDs, and a column of the number of reads that the tool counted in that particular gene.  

Each sample has data in a different subfolder in "../Datasets/RNASeqData", but each file we want has "htseq.txt" on the end of the file name.  Here we are going to go find a list of those files, and then next read each of them into R and store them as a list of data frames.  After we name each data frame in this list, we will merge them all into one large data frame. We provide an example of saving this resulting counts table in the repo and reading it back in for convenience for your later work.   
```{r}
htseqFiles <- list.files(path = "../Datasets/RNASeqData", # the path to where the data are saved
                        recursive = TRUE, # look inside directories that are in the path above
                        pattern = "htseq.txt", # only return results that have filenames that include this string
                        full.names = TRUE) # give back the full path/names so we can find where they are!

htseqNames <- gsub("-R....-.*", "", gsub("../Datasets/RNASeqData/", "", htseqFiles)) # extracting the sample name from the path!

dataList <- lapply(htseqFiles, function(x) {
  Y <- read.delim(file = x,
             header = FALSE,
             sep = "\t");
  colnames(Y) <- c("Gene", "Count")
  Y
})
names(dataList) <- htseqNames

studyData <- purrr::map_dfr(dataList, cbind, .id = "omics_sample_name")

studyCountsTable <- tidyr::spread(data = studyData, key = "omics_sample_name", value = "Count")
```

Write this counts table to a file for future use.  
```{r eval=FALSE, include=FALSE}
write.csv(x = studyCountsTable,
          file = "Results/studyCountsTable.csv",
          quote = F,
          row.names = F)
```

Read in this counts table in the future.  
```{r eval=FALSE, include=FALSE}
studyCountsTable <- read.csv(file = "Results/studyCountsTable.csv",
                             header = T,
                             stringsAsFactors = F)
```

### Explore your RNA Seq data
How many genes are there?
```{r}
length(unique(studyCountsTable$Gene))
```

How many total read counts did we get for each sample?
```{r}
readCounts <- colSums(studyCountsTable[,-1]) # column sum for all except the first Gene column
hist(readCounts)
max(readCounts)
min(readCounts)
```




## Creating Analysis Ready Data
Next you will need to create an analysis-ready dataset ready for analysis via DESeq2 in order to get differential gene expression results. 
```{r}



```




